### PR TITLE
DRAFT - feat: Add local external file refs

### PR DESCRIPTION
**Pull request by dpwdec**
## About the PR
Currently, referenced objects in the AsyncAPI spec when read by the AsyncAPI.NET library can only be parsed using internal document references to the `components` object. While this is very useful, ideally we would utilise the full `$ref` functionality afforded to us by referencing external files, both locally and remotely. The advantages of facilitating references outside of a single AsyncAPI file are myriad, but most importantly it:

- allows for re-use of referenced objects between AsyncAPI service files, for example, two services wanting to consume the same message can directly `$ref` a producer's definition of their message without need for duplication
- allows for AsyncAPI files to be defined modularly according to users needs as projects defined using AsyncAPI scale

This PR aims to implement the functionality for allowing referencing of *local* references, i.e. other AsyncAPI objects defined in YAML that exist on the same local file system.

### Challenges to solve

- How can we facilitate allowing users to pass in their own file reader as a settings option so as to make the resolving of file paths agnostic within the project itself?
- How can we propagate settings of readers down to external readers?
- Do we need to generalise the patterning that we use for defining external references to lower code duplication?

### Changelog
TODO

### Related Issues
TODO

### CODE REVIEW

File: src/LEGO.AsyncAPI.Readers/V2/AsyncApiMessageDeserializer.cs [hunk 1]
```diff
@@ -94,7 +96,13 @@ public static AsyncApiMessage LoadMessage(ParseNode node)
             var pointer = mapNode.GetReferencePointer();
             if (pointer != null)
             {
-                return mapNode.GetReferencedObject<AsyncApiMessage>(ReferenceType.Message, pointer);
+                // pointer is an internal components reference:
+                if (pointer.StartsWith("#")) return mapNode.GetReferencedObject<AsyncApiMessage>(ReferenceType.Message, pointer);
+
+                // how can we pass in settings and options for user to inject their own file reader
+                // is version matching important?
```
Comment (VisualBean) [inline src/LEGO.AsyncAPI.Readers/V2/AsyncApiMessageDeserializer.cs:103]: Version matching becomes important once we need support for both V2 and V3 which is coming very soon. 
But important only on the read side, as we aren't looking to support writing to both version, but rather "upgrade" older versions to the newest. 

Does refs really allow for non-component refs? (External yes, but internal non-component refs?)

File: src/LEGO.AsyncAPI.Readers/V2/AsyncApiChannelBindingDeserializer.cs [hunk 2]
```diff
@@ -15,7 +17,13 @@ internal static AsyncApiBindings<IChannelBinding> LoadChannelBindings(ParseNode
             var pointer = mapNode.GetReferencePointer();
             if (pointer != null)
             {
-                return mapNode.GetReferencedObject<AsyncApiBindings<IChannelBinding>>(ReferenceType.ChannelBindings, pointer);
+                if (pointer.StartsWith("#"))
+                {
+                    return mapNode.GetReferencedObject<AsyncApiBindings<IChannelBinding>>(ReferenceType.ChannelBindings, pointer);
+                }
+
+                var referencedContent = File.ReadAllText(pointer);
```
[the remark this answers is not in the document]
Comment (dpwdec) [reply src/LEGO.AsyncAPI.Readers/V2/AsyncApiChannelBindingDeserializer.cs:25]: @VisualBean Will do. Thanks for the context!

[the remark this answers is not in the document]
Comment (VisualBean) [reply src/LEGO.AsyncAPI.Readers/V2/AsyncApiChannelBindingDeserializer.cs:25]: I think I have an idea of how to solve this actually, while providing the support for custom loaders (for http and similar)

[the remark this answers is not in the document]
Comment (VisualBean) [reply src/LEGO.AsyncAPI.Readers/V2/AsyncApiChannelBindingDeserializer.cs:25]: Leaving internal fragments aside (might be able to fix it) I think itll work with what i have in mind.

[the remark this answers is not in the document]
Comment (VisualBean) [reply src/LEGO.AsyncAPI.Readers/V2/AsyncApiChannelBindingDeserializer.cs:25]: @dpwdec  ive made some initial digging into this (just allowing any type of reference and correctly setting the values) without resolving anything (just marking them as unresolved to begin with).

I would also really appreciate if work could be done against the vnext branch, as this has some breaking changes (like changing from yamldotnet to system.text.json)

**Comment by VisualBean** [comment]
> How can we facilitate allowing users to pass in their own file reader as a settings option so as to make the resolving of file paths agnostic within the project itself?

Perhaps an IReferenceLoader could be used as a part of the `AsyncApiReaderSettings()` coupled with the ReferenceResolution enum (add another one for resolving 'all' references or whatever).

A basic ReferenceLoader could be made that fetches http and local fragments.

> Do we need to generalise the patterning that we use for defining external references to lower code duplication?

I believe i have solved this in another PR

**Pull request closed without merging**